platform/ios/src/MGLMapView.mm

```diff
@@ -1231,15 +1267,103 @@ - (void)setPreferredFramesPerSecond:(MGLMapViewPreferredFramesPerSecond)preferre
     [self updateDisplayLinkPreferredFramesPerSecond];
 }
 
+- (void)setEnablePresentsWithTransaction:(BOOL)enablePresentsWithTransaction
+{
+    if (_enablePresentsWithTransaction == enablePresentsWithTransaction)
+    {
+        return;
+    }
+    
+    _enablePresentsWithTransaction = enablePresentsWithTransaction;
+    
+    // If the map is visible, change the layer property too
+    if (self.window) {
+        CAEAGLLayer *eaglLayer = MGL_OBJC_DYNAMIC_CAST(_glView.layer, CAEAGLLayer);
+        eaglLayer.presentsWithTransaction = enablePresentsWithTransaction;
+    }
+}
+
+#ifdef MGL_RECREATE_GL_IN_AN_EMERGENCY
+// See https://github.com/mapbox/mapbox-gl-native/issues/14232
+- (void)emergencyRecreateGL {
+    MGLLogError(@"Rendering took too long - creating GL views");
+
+    CAEAGLLayer *eaglLayer = MGL_OBJC_DYNAMIC_CAST(_glView.layer, CAEAGLLayer);
+    eaglLayer.presentsWithTransaction = NO;
+
+    [self sleepGL:nil];
+
+    // Just performing a sleepGL/wakeGL pair isn't sufficient - in this case
+    // we can still get errors when calling bindDrawable. Here we completely
+    // recreate the GLKView
+    
+    [self.userLocationAnnotationView removeFromSuperview];
+    [_glView removeFromSuperview];
+    
+    _glView = [[GLKView alloc] initWithFrame:self.bounds context:_context];
+    _glView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
+    _glView.enableSetNeedsDisplay = NO;
+    _glView.drawableStencilFormat = GLKViewDrawableStencilFormat8;
+    _glView.drawableDepthFormat = GLKViewDrawableDepthFormat16;
+    _glView.contentScaleFactor = [UIScreen instancesRespondToSelector:@selector(nativeScale)] ? [[UIScreen mainScreen] nativeScale] : [[UIScreen mainScreen] scale];
+    _glView.layer.opaque = _opaque;
+    _glView.delegate = self;
```

Comment:
These several lines are copy-pasted from `-createGLView`. Can we refactor that method to avoid the duplication? I’m just concerned that we may find it useful to keep this method around longer-term, but it’s really easy to forget to keep two code paths in sync.